cmd/test-server/main.go

```diff
@@ -178,48 +196,164 @@ func pingHandler(w http.ResponseWriter, r *http.Request) {
 		panic(err)
 	}
 	stat := pinger.Statistics()
-	var icmp []RtItem
-	icmp = append(icmp, RtItem{clientIP, stat.PacketsSent, stat.PacketsRecv, stat.PacketLoss, fmtTimeMs(stat.MinRtt), fmtTimeMs(stat.AvgRtt), fmtTimeMs(stat.MaxRtt), fmtTimeMs(stat.StdDevRtt)})
+	icmp := RtItem{ip, stat.PacketsSent, stat.PacketsRecv, stat.PacketLoss, fmtTimeMs(stat.MinRtt), fmtTimeMs(stat.AvgRtt), fmtTimeMs(stat.MaxRtt), fmtTimeMs(stat.StdDevRtt)}
+	return icmp
+}
 
+func TcpPinger(ip string) tcpStruct {
 	// TCP Pinger
 	var tcpResultArr []tcpResult
 	rand.Seed(time.Now().UnixNano()) // Or each time we restart server the sequences would repeat
+	var portsWaitGroup sync.WaitGroup
+	portsWaitGroup.Add(len(PortsToTest))
 	for _, port := range PortsToTest {
-		var seqNumber uint64 = uint64(rand.Uint32())
-		var dst = fmt.Sprintf("%s:%d", clientIP, port)
-		ticker := time.NewTicker(TCPInterval)
-		var tResult []float64
-		for x := 0; x < TCPCounter; x++ {
-			seqNumber++
-			select {
-			case <-ticker.C:
-				tResult = append(tResult, pingTcp(dst, seqNumber, TCPTimeout))
+		go func(port int) {
+			defer portsWaitGroup.Done()
+			var seqNumber uint64 = uint64(rand.Uint32())
+			var dst = fmt.Sprintf("%s:%d", ip, port)
+			ticker := time.NewTicker(TCPInterval)
+			var tResult []float64
+			for x := 0; x < TCPCounter; x++ {
+				seqNumber++
+				select {
+				case <-ticker.C:
+					tResult = append(tResult, sendTcpPing(dst, seqNumber, TCPTimeout))
+				}
+			}
+			ticker.Stop()
+			min, avg, max, stddev := getStats(tResult)
+			tcpResultArr = append(tcpResultArr, tcpResult{port, tResult, min, avg, max, stddev})
+		}(port)
+	}
+	portsWaitGroup.Wait()
+	tcpResultObj := tcpStruct{ip, tcpResultArr}
+	return tcpResultObj
+}
+
+// Avg RTT from all successful ICMP measurements, to display on webpage
+func getAvgIcmpResult(icmp []RtItem) float64 {
+	var sum float64 = 0
+	var len float64 = 0
+	for _, x := range icmp {
+		if x.AvgRtt == 0 {
+			continue
+		}
+		sum += x.AvgRtt
+		len += 1
+	}
+	var avg float64 = sum / len
+	return avg
+}
+
+// Avg RTT from all successful TCP measurements regardless of port, to display on webpage
+func getAvgTcpResult(tcp []tcpStruct) float64 {
+	var sum float64 = 0
+	var len float64 = 0
+	// for each IP in subnet
+	for _, x := range tcp {
+		// for each port per IP
+		for _, p := range x.Probes {
+			if p.AvgRtt == 0 {
+				continue
 			}
+			sum += p.AvgRtt
+			len += 1
+		}
+	}
+	var avg float64 = sum / len
+	return avg
+}
+
+// Handler for ICMP and TCP measurements which also serves the webpage via a template
+func pingHandler(w http.ResponseWriter, r *http.Request) {
```

Comment:
Nitpick: This is a fairly complex function that would benefit from being split into smaller functions. For example, one function could check for the correct HTTP parameters (lines 269-277); another could take care of the pinging loop (lines 288-316); and another one could take care of turning the result into JSON. A nice side effect of smaller functions is that they become much easier to test. In its current form, you'd have to bend over backwards to get good test coverage for this function. Anyway, this is fine for now but I figured I would elaborate on my thinking process!

Reply:
Ah thanks for the detail. Yeah, for now I've refactored it to take out the HTTP parameter checking because I do that for all three handlers. I'm leaving the pinging looping as it is because that is subject to change but I see your point about making it easier to read and to test. Thanks! 👍